blockchain/chain/src/store/chain_store.rs

```diff
@@ -471,19 +472,25 @@ where
             tpow = state.total_quality_adj_power;
         }
     }
-    let log2_p = if tpow > BigUint::zero() {
-        BigUint::from(tpow.bits() - 1)
+    let log2_p = if tpow > BigInt::zero() {
+        BigInt::from(tpow.bits() - 1)
     } else {
         return Err(
             "All power in the net is gone. You network might be disconnected, or the net is dead!"
                 .to_owned(),
         );
     };
 
-    let mut out = ts.weight() + (&log2_p << 8);
-    let e_weight = ((log2_p * BigUint::from(ts.blocks().len())) * BigUint::from(W_RATIO_NUM)) << 8;
-    let value = e_weight / (BigUint::from(BLOCKS_PER_EPOCH) * BigUint::from(W_RATIO_DEN));
-    out += &value;
+    let out_add: BigInt = &log2_p << 8;
+    let mut out = ts.weight()
+        + out_add
+            .to_biguint()
+            .ok_or_else(|| "Negative out".to_string())?;
```

Comment:
should just change the function signature to return a big int, no need for it to be unsigned (value is weight)